applications/RomApplication/custom_strategies/lspg_rom_builder_and_solver.h

```diff
@@ -0,0 +1,587 @@
+                    << "Reaction variable not set for the following :\n"
+                    << "Node : " << r_dof.Id() << '\n'
+                    << "Dof  : " << r_dof      << '\n'
+                    << "Not possible to calculate reactions." << std::endl;
+            }
+        }
+#endif
+        KRATOS_CATCH("");
+    } 
+    
+    void BuildAndSolve(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart &rModelPart,
+        TSystemMatrixType &A,
+        TSystemVectorType &Dx,
+        TSystemVectorType &b) override
+    {
+        KRATOS_TRY
+        LSPGSystemMatrixType Arom = ZeroMatrix(BaseType::GetEquationSystemSize(), this->GetNumberOfROMModes());
+        LSPGSystemVectorType brom = ZeroVector(BaseType::GetEquationSystemSize());
+        BuildROM(pScheme, rModelPart, Arom, brom);
+
+        // Obtain the assembled residuals vector (To build a basis for Petrov-Galerkin)
+        if (mTrainPetrovGalerkinFlag){
+            Kratos::Vector r_residual;
+            GetAssembledResiduals(pScheme, rModelPart, r_residual);
+        }
+
+        SolveROM(rModelPart, Arom, brom, Dx);
+
+
+        KRATOS_CATCH("")
+    }
+
+    Parameters GetDefaultParameters() const override
+    {
+        Parameters default_parameters = Parameters(R"(
+        {
+            "name" : "lspg_rom_builder_and_solver",
+            "nodal_unknowns" : [],
+            "number_of_rom_dofs" : 10,
+            "train_petrov_galerkin" : false
+        })");
+        default_parameters.AddMissingParameters(BaseType::GetDefaultParameters());
+
+        return default_parameters;
+    }
+
+    static std::string Name() 
+    {
+        return "lspg_rom_builder_and_solver";
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const override
+    {
+        return "LSPGROMBuilderAndSolver";
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream &rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    /// Print object's data.GetNumberOfROMModes()
+    virtual void PrintData(std::ostream &rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+
+    ///@}
+protected:
+
+    SizeType mNodalDofs;
+    bool mTrainPetrovGalerkinFlag = false;
+
+    ///@}
+    ///@name Protected static member variables
+    ///@{
+
+
+    ///@}
+    ///@name Protected member variables
+    ///@{
+
+    ///@}
+    ///@name Protected operators
+    ///@{
+
+
+    ///@}
+    ///@name Protected operations
+    ///@{
+
+    void AssignSettings(const Parameters ThisParameters) override
+    {
+        BaseType::AssignSettings(ThisParameters);
+
+        // // Set member variables
+        mTrainPetrovGalerkinFlag = ThisParameters["train_petrov_galerkin"].GetBool();

```

Comment:
petrov-Galerkin or LSPG here?

Reply:
Petrov-Galerkin is correct. Petrov-Galerkin is trained with the LSPG B&S.